File: src/lib/engine/ai-cast.ts
import { requestBrain } from "@/lib/client/api";
import type { TTSProviderMeta } from "@/lib/providers";
import type { Analysis, CastEntry, VoiceProfile } from "@/lib/types";
import { NARRATOR_ID } from "@/lib/types";
import { clamp, extractJSON, withRetry } from "@/lib/utils";
import { autoCast, candidateVoices, supportsPitch } from "./casting";
import { languageName } from "./lang";

export interface CastOutcome {
  cast: Record<string, CastEntry>;
  reasons: Record<string, string>;
  by: "ai" | "rules";
  note?: string;
}

interface CastOptions {
  analysis: Analysis;
  voices: VoiceProfile[];
  meta: TTSProviderMeta;
  model: string;
  signal?: AbortSignal;
}

interface RawPick {
  voice?: unknown;
  pitch?: unknown;
  rate?: unknown;
  why?: unknown;
}

const MAX_VOICES = 90;
const FALLBACK_NOTE = "The director couldn't finish casting this time, so voices were matched automatically.";

const SYSTEM =
  "You are the voice casting director of a professional audiobook studio. You match every speaker of a story to the single best voice from a voice catalogue, judging gender, age, personality, speaking style and the story's language, and you keep every main character clearly distinguishable. You always answer with a single JSON object and nothing else.";

function voiceLine(v: VoiceProfile) {
  const who = [v.gender, v.age].filter(Boolean).join(", ");
  const langs = v.langs.includes("*") ? "any language" : v.langs.join("/");
  return `${v.id} | ${v.name} | ${who} | ${langs}${v.multilingual ? " (multilingual)" : ""} | ${v.tags.slice(0, 5).join(", ")}`;
}

function buildPrompt(analysis: Analysis, voices: VoiceProfile[], pitchOK: boolean, engine: string) {
  const speakers = [
    `- narrator: ${analysis.narrator.gender} narrator; tone: ${analysis.narrator.tone}; ${analysis.pov}-person story`,
    ...[...analysis.characters]
      .sort((a, b) => b.lineCount - a.lineCount)
      .map(
        (c) =>
          `- ${c.id}: ${c.name} — ${c.gender}, ${c.age}, ${c.role}, ${c.lineCount} lines; personality: ${
            c.personality.join(", ") || "unknown"
          }; speaks: ${c.speakingStyle || "n/a"}; ideal voice: ${c.voiceDescription || "n/a"}`,
      ),
  ].join("\n");

  return `Cast the audiobook "${analysis.title}" (${languageName(analysis.language)}) with the ${engine} voice engine.

Return JSON: {"cast": {"narrator": {"voice": "voice id", "pitch": 0, "rate": 0, "why": "one short sentence"}, "c1": {"voice": "...", "pitch": 0, "rate": 0, "why": "..."}}}
Include every speaker listed below, keyed by its id.

Rules:
1. Use only voice ids from the catalogue, copied exactly.
2. Match gender first, then age: youthful voices for children and teens, mature voices for elders.
3. Give every main character a different voice. Reuse a voice only for minor characters, and then shift pitch or rate so they still sound distinct.
4. The narrator needs a clear, steady storytelling voice that is different from the lead characters.
5. Prefer voices native to the story's language; use multilingual voices when needed.
6. "rate" is a pace offset in percent from -30 to 30 that fits how the speaker talks: slow, wise or menacing below 0; quick, excitable or young above 0.
7. "pitch" is an offset in percent from -20 to 20${pitchOK ? "" : ". This engine cannot shift pitch, so always use 0"}.
8. "why" explains the choice in one short sentence about the voice, without naming any software.

SPEAKERS:
${speakers}

VOICE CATALOGUE (id | name | gender, age | languages | traits):
${voices.map(voiceLine).join("\n")}`;
}

function keepDistinct(cast: Record<string, CastEntry>, analysis: Analysis, pitchOK: boolean) {
  const nudges = [0, 8, -8, 14, -14, 5, -5];
  const seen = new Map<string, number>();
  const order = [NARRATOR_ID, ...[...analysis.characters].sort((a, b) => b.lineCount - a.lineCount).map((c) => c.id)];
  for (const id of order) {
    const entry = cast[id];
    if (!entry) continue;
    const n = seen.get(entry.voiceId) ?? 0;
    seen.set(entry.voiceId, n + 1);
    if (n === 0) continue;
    const k = nudges[n % nudges.length];
    cast[id] = {
      ...entry,
      pitch: pitchOK ? clamp(entry.pitch + k, -30, 30) : entry.pitch,
      rate: clamp(entry.rate + (pitchOK ? k / 2 : k), -35, 35),
    };
  }
}

export async function castVoices(o: CastOptions): Promise<CastOutcome> {
  const rules = autoCast(o.analysis, o.voices, o.meta, o.model);
  const pool = candidateVoices(o.voices, o.meta, o.analysis.language);
  const catalogue = [...pool.filter((p) => p.native), ...pool.filter((p) => !p.native)]
    .slice(0, MAX_VOICES)
    .map((p) => p.voice);
  if (!catalogue.length) return { cast: rules, reasons: {}, by: "rules" };
  const pitchOK = supportsPitch(o.meta.id, o.model);
  const exact = new Set(catalogue.map((v) => v.id));
  const loose = new Map(catalogue.map((v) => [v.id.toLowerCase(), v.id]));

  try {
    const raw = await withRetry(
      () => requestBrain(SYSTEM, buildPrompt(o.analysis, catalogue, pitchOK, o.meta.name), { signal: o.signal }),
      2,
    );
    const parsed = extractJSON<{ cast?: Record<string, RawPick> } & Record<string, RawPick>>(raw);
    const picks: Record<string, RawPick> = parsed.cast ?? parsed;
    const speakerIds = [NARRATOR_ID, ...o.analysis.characters.map((c) => c.id)];
    const cast = { ...rules };
    const reasons: Record<string, string> = {};
    let hits = 0;

    for (const id of speakerIds) {
      const pick = picks[id];
      if (!pick || typeof pick.voice !== "string") continue;
      const voiceId = exact.has(pick.voice) ? pick.voice : loose.get(pick.voice.trim().toLowerCase());
      if (!voiceId) continue;
      hits++;
      cast[id] = {
        voiceId,
        pitch: pitchOK ? clamp(Number(pick.pitch) || 0, -30, 30) : 0,
        rate: clamp(Number(pick.rate) || 0, -35, 35),
      };
      if (typeof pick.why === "string" && pick.why.trim()) reasons[id] = pick.why.trim().slice(0, 220);
    }

    if (hits < Math.ceil(speakerIds.length / 2)) return { cast: rules, reasons: {}, by: "rules", note: FALLBACK_NOTE };
    keepDistinct(cast, o.analysis, pitchOK);
    return { cast, reasons, by: "ai" };
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") throw err;
    if ((err as { status?: number }).status === 401) throw err;
    return { cast: rules, reasons: {}, by: "rules", note: FALLBACK_NOTE };
  }
}

File: android/app/src/main/java/com/nigixmosha/app/engine/AiCast.java
package com.nigixmosha.app.engine;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.nigixmosha.app.engine.model.Types;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class AiCast {
    public interface Brain {
        String ask(String system, String prompt) throws Exception;
    }

    public static final class Outcome {
        public Map<String, Types.CastEntry> cast;
        public Map<String, String> reasons = new LinkedHashMap<>();
        public String by;
        public String note;
    }

    private static final int MAX_VOICES = 90;
    private static final String FALLBACK_NOTE = "The director couldn't finish casting this time, so voices were matched automatically.";
    private static final String SYSTEM =
            "You are the voice casting director of a professional audiobook studio. You match every speaker of a story to the single best voice from a voice catalogue, judging gender, age, personality, speaking style and the story's language, and you keep every main character clearly distinguishable. You always answer with a single JSON object and nothing else.";

    private AiCast() {
    }

    private static String voiceLine(Types.VoiceProfile v) {
        List<String> who = new ArrayList<>();
        if (v.gender != null && !v.gender.isEmpty()) who.add(v.gender);
        if (v.age != null && !v.age.isEmpty()) who.add(v.age);
        String langs = v.langs.contains("*") ? "any language" : String.join("/", v.langs);
        List<String> tags = v.tags.subList(0, Math.min(5, v.tags.size()));
        return v.id + " | " + v.name + " | " + String.join(", ", who) + " | " + langs
                + (v.isMultilingual() ? " (multilingual)" : "") + " | " + String.join(", ", tags);
    }

    private static String prompt(Types.Analysis a, List<Types.VoiceProfile> voices, boolean pitchOK, String engine) {
        StringBuilder speakers = new StringBuilder();
        speakers.append("- narrator: ").append(a.narrator.gender).append(" narrator; tone: ").append(a.narrator.tone)
                .append("; ").append(a.pov).append("-person story");
        for (Types.CastCharacter c : Casting.byLines(a)) {
            speakers.append("\n- ").append(c.id).append(": ").append(c.name).append(" — ").append(c.gender).append(", ")
                    .append(c.age).append(", ").append(c.role).append(", ").append(c.lineCount).append(" lines; personality: ")
                    .append(c.personality.isEmpty() ? "unknown" : String.join(", ", c.personality))
                    .append("; speaks: ").append(c.speakingStyle == null || c.speakingStyle.isEmpty() ? "n/a" : c.speakingStyle)
                    .append("; ideal voice: ").append(c.voiceDescription == null || c.voiceDescription.isEmpty() ? "n/a" : c.voiceDescription);
        }
        StringBuilder catalogue = new StringBuilder();
        for (int i = 0; i < voices.size(); i++) {
            if (i > 0) catalogue.append('\n');
            catalogue.append(voiceLine(voices.get(i)));
        }
        return "Cast the audiobook \"" + a.title + "\" (" + Lang.languageName(a.language) + ") with the " + engine + " voice engine.\n\n"
                + "Return JSON: {\"cast\": {\"narrator\": {\"voice\": \"voice id\", \"pitch\": 0, \"rate\": 0, \"why\": \"one short sentence\"}, \"c1\": {\"voice\": \"...\", \"pitch\": 0, \"rate\": 0, \"why\": \"...\"}}}\n"
                + "Include every speaker listed below, keyed by its id.\n\n"
                + "Rules:\n"
                + "1. Use only voice ids from the catalogue, copied exactly.\n"
                + "2. Match gender first, then age: youthful voices for children and teens, mature voices for elders.\n"
                + "3. Give every main character a different voice. Reuse a voice only for minor characters, and then shift pitch or rate so they still sound distinct.\n"
                + "4. The narrator needs a clear, steady storytelling voice that is different from the lead characters.\n"
                + "5. Prefer voices native to the story's language; use multilingual voices when needed.\n"
                + "6. \"rate\" is a pace offset in percent from -30 to 30 that fits how the speaker talks: slow, wise or menacing below 0; quick, excitable or young above 0.\n"
                + "7. \"pitch\" is an offset in percent from -20 to 20" + (pitchOK ? "" : ". This engine cannot shift pitch, so always use 0") + ".\n"
                + "8. \"why\" explains the choice in one short sentence about the voice, without naming any software.\n\n"
                + "SPEAKERS:\n" + speakers + "\n\n"
                + "VOICE CATALOGUE (id | name | gender, age | languages | traits):\n" + catalogue;
    }

    private static void keepDistinct(Map<String, Types.CastEntry> cast, Types.Analysis a, boolean pitchOK) {
        int[] nudges = {0, 8, -8, 14, -14, 5, -5};
        Map<String, Integer> seen = new HashMap<>();
        List<String> order = new ArrayList<>();
        order.add(Types.NARRATOR_ID);
        for (Types.CastCharacter c : Casting.byLines(a)) order.add(c.id);
        for (String id : order) {
            Types.CastEntry entry = cast.get(id);
            if (entry == null) continue;
            int n = seen.getOrDefault(entry.voiceId, 0);
            seen.put(entry.voiceId, n + 1);
            if (n == 0) continue;
            int k = nudges[n % nudges.length];
            cast.put(id, new Types.CastEntry(entry.voiceId,
                    pitchOK ? Util.clamp(entry.pitch + k, -30, 30) : entry.pitch,
                    Util.clamp(entry.rate + (pitchOK ? k / 2.0 : k), -35, 35)));
        }
    }

    private static double number(JsonElement el) {
        try {
            return el == null || el.isJsonNull() ? 0 : el.getAsDouble();
        } catch (Exception e) {
            return 0;
        }
    }

    public static Outcome castVoices(Types.Analysis analysis, List<Types.VoiceProfile> voices, Types.ProviderMeta meta,
                                     String model, Brain brain, Cancel cancel) throws Exception {
        Map<String, Types.CastEntry> rules = Casting.autoCast(analysis, voices, meta, model);
        List<Casting.Candidate> pool = Casting.candidateVoices(voices, meta, analysis.language);
        List<Types.VoiceProfile> catalogue = new ArrayList<>();
        for (Casting.Candidate c : pool) if (c.nativeVoice) catalogue.add(c.voice);
        for (Casting.Candidate c : pool) if (!c.nativeVoice) catalogue.add(c.voice);
        if (catalogue.size() > MAX_VOICES) catalogue = new ArrayList<>(catalogue.subList(0, MAX_VOICES));
        Outcome outcome = new Outcome();
        if (catalogue.isEmpty()) {
            outcome.cast = rules;
            outcome.by = "rules";
            return outcome;
        }
        boolean pitchOK = Casting.supportsPitch(meta.id, model);
        Set<String> exact = new HashSet<>();
        Map<String, String> loose = new HashMap<>();
        for (Types.VoiceProfile v : catalogue) {
            exact.add(v.id);
            loose.put(v.id.toLowerCase(java.util.Locale.ROOT), v.id);
        }
        List<Types.VoiceProfile> finalCatalogue = catalogue;
        try {
            String raw = Util.withRetry(() -> brain.ask(SYSTEM, prompt(analysis, finalCatalogue, pitchOK, meta.name)), 2, cancel);
            JsonElement parsed = Util.extractJSON(raw);
            JsonObject root = parsed.isJsonObject() ? parsed.getAsJsonObject() : new JsonObject();
            JsonObject picks = root.has("cast") && root.get("cast").isJsonObject() ? root.getAsJsonObject("cast") : root;
            List<String> speakerIds = new ArrayList<>();
            speakerIds.add(Types.NARRATOR_ID);
            for (Types.CastCharacter c : analysis.characters) speakerIds.add(c.id);
            Map<String, Types.CastEntry> cast = new LinkedHashMap<>(rules);
            Map<String, String> reasons = new LinkedHashMap<>();
            int hits = 0;
            for (String id : speakerIds) {
                JsonElement pickEl = picks.get(id);
                if (pickEl == null || !pickEl.isJsonObject()) continue;
                JsonObject pick = pickEl.getAsJsonObject();
                JsonElement voiceEl = pick.get("voice");
                if (voiceEl == null || !voiceEl.isJsonPrimitive() || !voiceEl.getAsJsonPrimitive().isString()) continue;
                String voice = voiceEl.getAsString();
                String voiceId = exact.contains(voice) ? voice : loose.get(voice.trim().toLowerCase(java.util.Locale.ROOT));
                if (voiceId == null) continue;
                hits++;
                cast.put(id, new Types.CastEntry(voiceId,
                        pitchOK ? Util.clamp(number(pick.get("pitch")), -30, 30) : 0,
                        Util.clamp(number(pick.get("rate")), -35, 35)));
                String why = Util.str(pick.get("why"));
                if (!why.isEmpty()) reasons.put(id, why.length() > 220 ? why.substring(0, 220) : why);
            }
            if (hits < Math.ceil(speakerIds.size() / 2.0)) {
                outcome.cast = rules;
                outcome.by = "rules";
                outcome.note = FALLBACK_NOTE;
                return outcome;
            }
            keepDistinct(cast, analysis, pitchOK);
            outcome.cast = cast;
            outcome.reasons = reasons;
            outcome.by = "ai";
            return outcome;
        } catch (Cancel.Cancelled c) {
            throw c;
        } catch (Exception e) {
            if (Util.status(e) == 401) throw e;
            outcome.cast = rules;
            outcome.by = "rules";
            outcome.note = FALLBACK_NOTE;
            return outcome;
        }
    }
}
